source/Halibut/Transport/Protocol/MessageExchangeStream.cs

```diff
@@ -294,15 +294,69 @@ static DataStream FindStreamById(StreamCapture capture, Guid id)
 
         void WriteBsonMessage<T>(T messages)
         {
-            using (var buffer = new BufferedStream(stream, 4096, true))
-            using (var zip = new DeflateStream(buffer, CompressionMode.Compress, true))
-            using (var bson = new BsonDataWriter(zip) { CloseOutput = false })
+            using (var zip = new DeflateStream(stream, CompressionMode.Compress, true))
+            using (var buffer = new BufferedStream(new NonDisposableStreamWrapper(zip), 4096))
+            using (var bson = new BsonDataWriter(buffer) { CloseOutput = false })
             {
                 serializer.Serialize(bson, new MessageEnvelope { Message = messages });
                 bson.Flush();
             }
         }
 
+        class NonDisposableStreamWrapper : Stream
```

Comment:
Why is this needed?

Reply:
We want to leave the wrapped stream opened.

Reply:
At least that is what the previous code was doing

Reply:
But maybe given we explicitly saying to leave the inner stream open, we don't need it!

Reply:
Ok I have removed NonDisposableStreamWrapper

Reply:
The original code passes `true` to the `BufferedStream` ctor while the new code doesn't.

Reply:
It looks like Halibut has it's own `BufferedStream` implementation

Reply:
Ah, you removed it, GH hid that change away from me :)

Reply:
The `DeflateStream` can keep the stream open, and that's passed into the `BufferedStream` so we don't need a custom implementation for that anymore.

Reply:
👍 Makes sense,only the first stream in the stack needs a keep open